File: XamlTag/Xaml.fs
namespace XamlTag

open System
open System.Collections.Generic
open System.Dynamic

type Xaml<'a> internal (b : IXamlBuilder, c : SetterFactory)=
  inherit DynamicObject()
  
  let model = new ConstructModel<'a>(b,c)
  let thing = lazy(Activator.CreateInstance<'a>() |> model.Play)
  
  let (|Single|Multi|NestedFunc|NestedManyFunc|) (binder : InvokeMemberBinder, args : Object[]) =
    if binder.Name.Contains("And") then
      Multi
    elif args.[0] :? Func<IXamlBuilder,XamlCreator> then
      NestedFunc
    elif args.[0] :? Func<IXamlBuilder,XamlCreator[]> then
      NestedManyFunc
    else
      Single
     
  member x.Create()=thing.Force()

  interface XamlCreator with
    member x.GetXamlObject()=
      x.Create() :> Object

  override x.TryInvokeMember(binder, args, result)=
    match (binder,args) with
    | Single -> model.AddSingle binder.Name args.[0]
    | Multi ->  model.AddMulti binder.Name args
    | NestedFunc -> model.AddNested binder.Name args
    | NestedManyFunc -> model.AddNestedMany binder.Name args
    result <- x
    true
      
    

File: DynamicXaml/Xaml.cs
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using DynamicXaml.Extensions;

namespace DynamicXaml
{
    public interface Xaml
    {
        object Create();
    }

    public class Xaml<T> : DynamicObject, Xaml
    {
        private readonly XamlBuilder _xamlBuilder;
        private readonly object _dataContext;
        private readonly List<InvokeMemberHandler> _invokeMemberHandler;
        private readonly CreationModel<T> _creationModel;
        private readonly Lazy<T> _created;

        public Xaml(XamlBuilder xamlBuilder,  object dataContext = null)
        {
            _xamlBuilder = xamlBuilder;
            _dataContext = dataContext;
            _creationModel = new CreationModel<T>(dataContext);
            _invokeMemberHandler = _xamlBuilder.GetInvokeMemberHandler().ToList();
            _created = new Lazy<T>(()=>_creationModel.Play(Activator.CreateInstance<T>()));
        }

        public override bool TryInvokeMember(InvokeMemberBinder binder, object[] args, out object result)
        {
            var callContext = new RootInvokeContext<T>(binder, args, _xamlBuilder, _invokeMemberHandler);
            callContext.TransferRecordedActionsInto(_creationModel);
            result = this;
            return true;
        }

        public T Create()
        {
            return _created.Value;
        }

        object Xaml.Create()
        {
            return Create();
        }
    }
}